sycl/source/half_type.cpp

```diff
@@ -105,7 +106,9 @@ static float half2Float(const uint16_t &Val) {
   Bits |= (Exp32 << 23);
   Bits |= Frac32;
 
-  return *reinterpret_cast<float *>(&Bits);
+  float Result;
```

Comment:
I think most compiler will be able to optimize away this memcpy call, but how about this fix:
```c++
union {
  float Result;
  uint32_t Bits;
} U;

U.Bits = Sign | (Exp32 << 23) | Frac32;
return U.Result;
```

Reply:
Strictly speaking, this implicit reinterpret cast between union fields is a UB by the C++ spec. From [here](https://en.cppreference.com/w/cpp/language/union):

> The details of that allocation are implementation-defined, and **it's undefined behavior to read from the member of the union that wasn't most recently written**. Many compilers implement, as a non-standard language extension, the ability to read inactive members of a union.

Reply:
I think no need to change one UB to another.